sw/host/opentitanlib/src/uart/console.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use regex::{Captures, Regex};
+use std::fs::File;
+use std::io::{Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::time::{Duration, Instant, SystemTime};
+
+use crate::io::uart::Uart;
+use crate::util::file;
+
+#[derive(Default)]
+pub struct UartConsole {
+    pub logfile: Option<File>,
+    pub timeout: Option<Duration>,
+    pub deadline: Option<Instant>,
+    pub exit_success: Option<Regex>,
+    pub exit_failure: Option<Regex>,
+    pub timestamp: bool,
+    pub buffer: String,
+    pub newline: bool,
+}
+
+#[derive(Clone, Copy)]
+pub enum ExitStatus {
+    None,
+    CtrlC,
+    Timeout,
+    ExitSuccess,
+    ExitFailure,
+}
+
+pub trait ReadAsRawFd: Read + AsRawFd {}
+
+impl<T: Read + AsRawFd> ReadAsRawFd for T {}
+
+impl UartConsole {
+    const CTRL_C: u8 = 3;
+    const BUFFER_LEN: usize = 1024;
+
+    // Runs an interactive console until CTRL_C is received.
+    pub fn interact(
+        &mut self,
+        uart: &dyn Uart,
+        mut stdin: Option<&mut dyn ReadAsRawFd>,
+        mut stdout: Option<&mut dyn Write>,
+    ) -> Result<ExitStatus> {
+        if let Some(timeout) = &self.timeout {
```

Comment:
```rust
self.deadline = self.timeout.map(|tt| Instant::now() + tt)
```

Reply:
I only want to overwrite `deadline` with Some(value) if timeout is Some().  If timeout is None, I want deadline to preserve whatever it has (e.g. it may be Some(absolute time)).

Reply:
Got it. That wasn't super obvious.